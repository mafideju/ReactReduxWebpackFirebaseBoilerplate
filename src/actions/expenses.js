import uuid from 'uuid';
import database from '../firebase/firebase';

// FLUXO DE DADOS SEM FIREBASE
// 01: COMPONENTE CHAMA O ACTION GENERATOR
// 02: ACTION GENERATOR RETORNA UM OBJETO
// 03: COMPONENTE DISPACHA A AÇÃO ATRAVES DO MAP-DISPATCH-TO-PROPS
// 04: REDUX GUARDA AS MUDANÇAS

// FLUXO DE DADOS COM FIREBASE
// 01: COMPONENTE CHAMA O ACTION GENERATOR
// 02: ACTION GENERATOR RETORNA UMA FUNÇÃO
// 03: COMPONENTE DISPACHA FUNÇÃO (REDUX THUNK)
// 04: FUNÇÃO DISPARA UMA AÇÃO QUALQUER PREDETERMINADA

// 01: CRIAR AS ACTIONS PARA CADA AÇÃO DO USUÁRIO (COM IDENTIFICADOR UNICO)
// ADD_EXPENSE
export const addExpense = expense => ({
  type: 'ADD_EXPENSE',
  expense
});

// CONSIDERAÇÕES:
// 01: DATABASE UPDATE COM SUCESSO
// 02: DISPATCH O ACTION GENERATOR CORRETO
export const startAddExpense = (expenseData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData;
    const expense = { description, note, amount, createdAt };

    return database
      .ref(`users/${uid}/expenses`)
      .push(expense)
      .then(ref => {
        dispatch(
          addExpense({
            id: ref.key,
            ...expense
          })
        );
      });
  };
};

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
});

export const startRemoveExpense = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database
      .ref(`users/${uid}/expenses/${id}`)
      .remove()
      .then(() => {
        dispatch(removeExpense({ id }));
      });
  };
};

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

export const startEditExpense = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database
      .ref(`users/${uid}/expenses/${id}`)
      .update(updates)
      .then(() => {
        dispatch(editExpense(id, updates));
      });
  };
};

// SET_EXPENSES
export const setExpenses = expenses => ({
  type: 'SET_EXPENSES',
  expenses
});

export const startSetExpenses = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database
      .ref(`users/${uid}/expenses`)
      .once('value')
      .then(snapshot => {
        const expenses = [];
        snapshot.forEach(snap => {
          expenses.push({
            id: snap.key,
            ...snap.val()
          });
        });
        dispatch(setExpenses(expenses));
      });
  };
};
