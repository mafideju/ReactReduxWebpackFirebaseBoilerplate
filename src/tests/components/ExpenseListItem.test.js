import React from 'react';
import { shallow } from 'enzyme';
import expenses from '../fixture/expenses';
import ExpenseListItem from '../../components/ExpenseListItem';

test('EXPENSE LIST ITEM TEST SUITE', () => {
  const wrapper = shallow(<ExpenseListItem {...expenses[0]} />);
  expect(wrapper).toMatchSnapshot();
});
