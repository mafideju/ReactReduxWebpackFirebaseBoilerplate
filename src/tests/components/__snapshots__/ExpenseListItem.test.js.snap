// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EXPENSE LIST ITEM TEST SUITE 1`] = `
<div>
  <Link
    replace={false}
    to="/edit/1"
  >
    <h3>
      Aluguel
    </h3>
  </Link>
  <p>
    Valor: R$12.00 Reais
     -
     
    Vencimento: 31/12/1969
  </p>
</div>
`;
